File: src/main/java/xyz/pupbrained/config/DropConfirmConfig.java
package xyz.pupbrained.config;

import dev.isxander.yacl3.api.*;
import dev.isxander.yacl3.config.v2.api.ConfigClassHandler;
import dev.isxander.yacl3.config.v2.api.SerialEntry;
import dev.isxander.yacl3.config.v2.api.serializer.GsonConfigSerializerBuilder;
import dev.isxander.yacl3.gui.controllers.BooleanController;
import dev.isxander.yacl3.gui.controllers.dropdown.ItemController;
import dev.isxander.yacl3.gui.controllers.slider.DoubleSliderController;
import dev.isxander.yacl3.platform.YACLPlatform;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.item.Item;
import net.minecraft.item.Items;
import net.minecraft.text.Text;

import java.util.List;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

public final class DropConfirmConfig {
  public static final ConfigClassHandler<DropConfirmConfig> GSON = ConfigClassHandler.createBuilder(DropConfirmConfig.class)
    .serializer(config -> GsonConfigSerializerBuilder.create(config)
      .setPath(YACLPlatform.getConfigDir().resolve("drop_confirm.json"))
      .build())
    .build();

  @SerialEntry
  public boolean enabled = true;

  @SerialEntry
  public boolean playSounds = true;

  @SerialEntry
  public boolean treatAsWhitelist = false;

  @SerialEntry
  public double confirmationResetDelay = 1.0;

  @SerialEntry
  public List<Item> blacklistedItems = List.of();

  public static Screen createScreen(Screen parent) {
    return YetAnotherConfigLib.create(DropConfirmConfig.GSON, ((defaults, config, builder) -> {
      var defaultCategoryBuilder = ConfigCategory.createBuilder()
        .name(Text.translatable("category.drop_confirm.general"));

      var enabled = createOption(
        "option.drop_confirm.enabled",
        "option.drop_confirm.enabled.description",
        defaults.enabled,
        () -> config.enabled,
        val -> config.enabled = val,
        booleanOption -> new BooleanController(booleanOption, true)
      );

      var playSounds = createOption(
        "option.drop_confirm.play_sounds",
        "option.drop_confirm.play_sounds.description",
        defaults.playSounds,
        () -> config.playSounds,
        val -> config.playSounds = val,
        booleanOption -> new BooleanController(booleanOption, true)
      );

      var treatAsWhitelist = createOption(
        "option.drop_confirm.treat_as_whitelist",
        "option.drop_confirm.treat_as_whitelist.description",
        defaults.treatAsWhitelist,
        () -> config.treatAsWhitelist,
        val -> config.treatAsWhitelist = val,
        booleanOption -> new BooleanController(booleanOption, true)
      );

      var confirmationResetDelay = createOption(
        "option.drop_confirm.confirmation_reset_delay",
        "option.drop_confirm.confirmation_reset_delay.description",
        defaults.confirmationResetDelay,
        () -> config.confirmationResetDelay,
        val -> config.confirmationResetDelay = val,
        doubleOption -> new DoubleSliderController(doubleOption, 1.0, 5.0, 0.05)
      );

      // FIXME: Figure out a way to update the strings instantly
      //        Currently, the strings are only updated when the
      //        config screen is reloaded.
      var blacklistedItems = createListOption(
        config.treatAsWhitelist ? "option.drop_confirm.whitelisted_items" : "option.drop_confirm.blacklisted_items",
        config.treatAsWhitelist ? "option.drop_confirm.whitelisted_items.description" : "option.drop_confirm.blacklisted_items.description",
        defaults.blacklistedItems,
        () -> config.blacklistedItems,
        val -> config.blacklistedItems = val,
        Items.AIR,
        true,
        ItemController::new
      );

      return builder
        .title(Text.translatable("config.drop_confirm.title"))
        .category(
          defaultCategoryBuilder
            .options(List.of(enabled, playSounds, confirmationResetDelay, treatAsWhitelist))
            .group(blacklistedItems)
            .build()
        );
    })).generateScreen(parent);
  }

  private static <T> Option<T> createOption (
    String name,
    String description,
    T defaultValue,
    Supplier<T> currentValue,
    Consumer<T> newValue,
    Function<Option<T>, Controller<T>> customController
  ) {
    return Option.<T>createBuilder()
      .name(Text.translatable(name))
      .description(
        OptionDescription.createBuilder()
          .text(Text.translatable(description))
          .build()
      )
      .binding(defaultValue, currentValue, newValue)
      .customController(customController)
      .build();
  }

  private static <T> ListOption<T> createListOption (
    String name,
    String description,
    List<T> defaultValue,
    Supplier<List<T>> currentValue,
    Consumer<List<T>> newValue,
    T initialValue,
    Boolean insertEntriesAtEnd,
    Function<ListOptionEntry<T>, Controller<T>> customController
  ) {
    return ListOption.<T>createBuilder()
      .name(Text.translatable(name))
      .description(
        OptionDescription.createBuilder()
          .text(Text.translatable(description))
          .build()
      )
      .binding(defaultValue, currentValue, newValue)
      .initial(initialValue)
      .insertEntriesAtEnd(insertEntriesAtEnd)
      .customController(customController)
      .build();
  }
}

File: common/src/main/kotlin/xyz/pupbrained/drop_confirm/DropConfirmConfig.kt
package xyz.pupbrained.drop_confirm

import dev.isxander.yacl3.api.*
import dev.isxander.yacl3.config.v2.api.ConfigClassHandler
import dev.isxander.yacl3.config.v2.api.SerialEntry
import dev.isxander.yacl3.config.v2.api.serializer.GsonConfigSerializerBuilder
import dev.isxander.yacl3.gui.controllers.BooleanController
import dev.isxander.yacl3.gui.controllers.dropdown.ItemController
import dev.isxander.yacl3.gui.controllers.slider.DoubleSliderController
import dev.isxander.yacl3.platform.YACLPlatform.getConfigDir
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.Screen
import net.minecraft.network.chat.Component
import net.minecraft.world.item.Item
import net.minecraft.world.item.Items

class DropConfirmConfig {
  @SerialEntry
  var enabled = true

  @SerialEntry
  var playSounds = true

  @SerialEntry
  var treatAsWhitelist = false

  @SerialEntry
  var confirmationResetDelay = 1.0

  @SerialEntry
  var blacklistedItems: List<Item> = emptyList()

  companion object {
    val GSON: ConfigClassHandler<DropConfirmConfig> = ConfigClassHandler.createBuilder(DropConfirmConfig::class.java)
      .serializer {
        GsonConfigSerializerBuilder.create(it)
          .setPath(getConfigDir().resolve("drop_confirm.json"))
          .setJson5(true)
          .build()
      }
      .build()

    @JvmStatic
    fun createScreen(parent: Screen): Screen = YetAnotherConfigLib.create(GSON) { defaults, config, builder ->
      fun String.t() = Component.translatable(this)
      fun String.desc() = OptionDescription.createBuilder().text(this.t()).build()

      fun <T : Any> option(
        key: String,
        default: T,
        get: () -> T,
        set: (T) -> Unit,
        controller: (Option<T>) -> Controller<T>,
      ): Option<T> = Option.createBuilder<T>()
        .name("option.drop_confirm.$key".t())
        .description("option.drop_confirm.$key.description".desc())
        .binding(default, get, set)
        .customController(controller)
        .build()

      val options = listOf(
        option(
          key = "enabled",
          default = defaults.enabled,
          get = { config.enabled },
          set = { config.enabled = it },
          controller = { BooleanController(it, true) }
        ),

        option(
          key = "play_sounds",
          default = defaults.playSounds,
          get = { config.playSounds },
          set = { config.playSounds = it },
          controller = { BooleanController(it, true) }
        ),

        option(
          key = "confirmation_reset_delay",
          default = defaults.confirmationResetDelay,
          get = { config.confirmationResetDelay },
          set = { config.confirmationResetDelay = it },
          controller = { DoubleSliderController(it, 1.0, 5.0, 0.05) }
        ),

        option(
          key = "treat_as_whitelist",
          default = defaults.treatAsWhitelist,
          get = { config.treatAsWhitelist },
          set = { value ->
            config.treatAsWhitelist = value
            Minecraft.getInstance().apply { execute { setScreen(createScreen(parent)) } }
          },
          controller = { BooleanController(it, true) }
        )
      )

      val listKey = if (config.treatAsWhitelist) "whitelisted" else "blacklisted"
      val itemList = ListOption.createBuilder<Item>().apply {
        name("option.drop_confirm.${listKey}_items".t())
        description("option.drop_confirm.${listKey}_items.description".desc())
        binding(defaults.blacklistedItems, { config.blacklistedItems }, { config.blacklistedItems = it })
        initial(Items.AIR)
        insertEntriesAtEnd(true)
        customController { ItemController(it) }
      }.build()

      builder.apply {
        title("config.drop_confirm.title".t())
        category(
          ConfigCategory.createBuilder().apply {
            name("category.drop_confirm.general".t())
            options(options)
            group(itemList)
          }.build()
        )
      }
    }.generateScreen(parent)
  }
}
